consensus/src/units/store.rs

```diff
@@ -1,291 +1,273 @@
-                        .map(|round| (id, round as Round)),
-                    None => None,
-                })
-                .collect(),
-        );
-        UnitStoreStatus::new(
-            node,
-            &self.is_forker,
-            self.by_coord.len(),
-            self.by_coord.keys().map(|k| k.round).max().unwrap_or(0),
-            top_row,
-            first_missing_rounds,
-        )
-    }
-
-    pub(crate) fn unit_by_coord(&self, coord: UnitCoord) -> Option<&SignedUnit<H, D, K>> {
-        self.by_coord.get(&coord)
+    fn mut_hashes_by(&mut self, creator: NodeIndex) -> &mut HashMap<Round, HashFor<U>> {
+        self.by_coord
+            .get_mut(creator)
+            .expect("all hashmaps initialized")
     }
 
-    pub(crate) fn unit_by_hash(&self, hash: &H::Hash) -> Option<&SignedUnit<H, D, K>> {
-        self.by_hash.get(hash)
-    }
-
-    pub(crate) fn contains_hash(&self, hash: &H::Hash) -> bool {
-        self.by_hash.contains_key(hash)
+    fn hashes_by(&self, creator: NodeIndex) -> &HashMap<Round, HashFor<U>> {
+        self.by_coord
+            .get(creator)
+            .expect("all hashmaps initialized")
     }
 
-    pub(crate) fn contains_coord(&self, coord: &UnitCoord) -> bool {
-        self.by_coord.contains_key(coord)
+    // only call this for canonical units
+    fn canonical_by_hash(&self, hash: &HashFor<U>) -> &U {
+        self.by_hash.get(hash).expect("we have all canonical units")
     }
 
-    pub(crate) fn newest_unit(
-        &self,
-        index: NodeIndex,
-    ) -> Option<UncheckedSignedUnit<H, D, K::Signature>> {
-        Some(
-            self.by_coord
-                .values()
-                .filter(|su| su.as_signable().creator() == index)
-                .max_by_key(|su| su.as_signable().round())?
-                .clone()
-                .into_unchecked(),
-        )
-    }
-
-    // Outputs new legit units that are supposed to be sent to Consensus and empties the buffer.
-    pub(crate) fn yield_buffer_units(&mut self) -> Vec<SignedUnit<H, D, K>> {
-        std::mem::take(&mut self.legit_buffer)
+    /// Insert a unit. If no other unit with this coord is in the store it becomes canonical.
+    pub fn insert(&mut self, unit: U) {
+        let unit_hash = unit.hash();
+        let unit_coord = unit.coord();
+        if self.canonical_unit(unit_coord).is_none() {
+            self.mut_hashes_by(unit_coord.creator())
+                .insert(unit.coord().round(), unit_hash);
+        }
+        self.by_hash.insert(unit_hash, unit);
     }
 
-    // Outputs None if this is not a newly-discovered fork or Some(sv) where (su, sv) form a fork
-    pub(crate) fn is_new_fork(&self, fu: &FullUnit<H, D>) -> Option<SignedUnit<H, D, K>> {
-        if self.contains_hash(&fu.hash()) {
-            return None;
+    /// Remove a unit with a given hash. Notably if you remove a unit another might become canonical in its place in the future.
+    pub fn remove(&mut self, hash: &HashFor<U>) {
```

Comment:
I don't see `remove` being used anywhere in the main code. Will it be used in the future?

Reply:
It's used within `validation` actually – in `finished_processing` iirc.

Reply:
Right, I missed that `processing_units` are of type `UnitStore`.